File: lintcode/unique-paths-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/unique-paths-ii/

class Solution:
    """
    @param obstacleGrid: An list of lists of integers
    @return: An integer
    """
    def uniquePathsWithObstacles(self, obstacleGrid):
        # write your code here
        m = len(obstacleGrid)
        if m == 0:
            return 0
        n = len(obstacleGrid[0])
        if n == 0:
            return 0

        matrix = [[0] * n for i in xrange(m)]
        for x in xrange(m):
            if obstacleGrid[x][0] == 1:
                matrix[x][0] = 0
            else:
                if x > 0:
                    matrix[x][0] = matrix[x-1][0]
                else:
                    matrix[x][0] = 1
        for y in xrange(n):
            if obstacleGrid[0][y] == 1:
                matrix[0][y] = 0
            else:
                if y > 0:
                    matrix[0][y] = matrix[0][y-1]
                else:
                    matrix[0][y] = 1

        for x in xrange(1, m):
            for y in xrange(1, n):
                if obstacleGrid[x][y] == 1:
                    matrix[x][y] = 0
                else:
                    matrix[x][y] = matrix[x-1][y] + matrix[x][y-1]

        return matrix[m-1][n-1]

if __name__ == '__main__':
    s = Solution()
    print s.uniquePathsWithObstacles([[0,0,0],[0,1,0],[0,0,0]])

File: leetcode-2/unique-paths-ii.cc
// Question: https://leetcode.com/problems/unique-paths-ii/

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        if (obstacleGrid.size() == 0 || obstacleGrid[0].size() == 0) {
            return 0;
        }

        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<vector<int>> paths(m, vector<int>(n, 0));
        paths[0][0] = (obstacleGrid[0][0] == 0 ? 1 : 0);
        for (int i = 1; i < m; ++i) {
            if (obstacleGrid[i][0] == 0) {
                paths[i][0] = paths[i-1][0];
            } else {
                paths[i][0] = 0;
            }
        }
        for (int j = 1; j < n; ++j) {
            if (obstacleGrid[0][j] == 0) {
                paths[0][j] = paths[0][j-1];
            } else {
                paths[0][j] = 0;
            }
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (obstacleGrid[i][j] == 0) {
                    paths[i][j] = paths[i-1][j] + paths[i][j-1];
                } else {
                    paths[i][j] = 0;
                }
            }
        }
        return paths.back().back();
    }
};
